rellic/AST/Z3ConvVisitor.cpp

```diff
@@ -869,38 +896,37 @@ void Z3ConvVisitor::VisitBinaryApp(z3::expr z_op) {
   clang::Expr *c_op{nullptr};
   switch (z_func.decl_kind()) {
     case Z3_OP_EQ:
-      c_op = CreateBinaryOperator(*ast_ctx, clang::BO_EQ, lhs, rhs,
-                                  ast_ctx->IntTy);
+      c_op = BinOpExpr(clang::BO_EQ, ast_ctx->IntTy);
       break;
 
     case Z3_OP_SLEQ:
-      c_op = CreateBinaryOperator(*ast_ctx, clang::BO_LE, lhs, rhs,
-                                  ast_ctx->IntTy);
+      c_op = BinOpExpr(clang::BO_LE, ast_ctx->IntTy);
       break;
 
     case Z3_OP_AND: {
-      c_op = lhs;
       for (auto i = 1U; i < z_op.num_args(); ++i) {
         rhs = GetCExpr(z_op.arg(i));
-        c_op = CreateBinaryOperator(*ast_ctx, clang::BO_LAnd, c_op, rhs,
-                                    ast_ctx->IntTy);
+        lhs = BinOpExpr(clang::BO_LAnd, ast_ctx->IntTy);
       }
+      c_op = lhs;
     } break;
 
     case Z3_OP_OR: {
-      c_op = lhs;
       for (auto i = 1U; i < z_op.num_args(); ++i) {
         rhs = GetCExpr(z_op.arg(i));
-        c_op = CreateBinaryOperator(*ast_ctx, clang::BO_LOr, c_op, rhs,
-                                    ast_ctx->IntTy);
+        lhs = BinOpExpr(clang::BO_LOr, ast_ctx->IntTy);
       }
+      c_op = lhs;
     } break;
 
     case Z3_OP_CONCAT: {
-      if (auto c_lit = clang::dyn_cast<clang::IntegerLiteral>(lhs)) {
+      auto c_lit{
+          clang::dyn_cast<clang::IntegerLiteral>(lhs->IgnoreParenCasts())};
+      if (c_lit && (c_lit->getValue().isNullValue() ||
+                    c_lit->getValue().isAllOnesValue())) {
         auto val = c_lit->getValue();
         auto s_size = GetZ3SortSize(z_op.get_sort());
-        auto t_sign = val.isNullValue() ? 0 : 1;
+        auto t_sign = val == 0 ? 0U : 1U;
```

Comment:
Can you add a comment here describing how signedness is inferred or communicated here?